### PR TITLE
Monaco Extend CT Streamlit application

### PR DESCRIPTION
**Pull request by SimonBiggs**
Add a Monaco CT extension streamlit app.

### CODE REVIEW

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (sjswerdloff) [inline lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:169]: What does this do?
I'm  on my phone and searching for the key function in a different place is awkward.  Does it duplicate the slice end but alter the image position patient by the difference of the previous slice (add it to the stack).
This is something people frequently get wrong because of assumptions

**Review by sjswerdloff — COMMENTED** [review]
Prefer responses before approval

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:169]: This function being called is code from 2019 which is within the following:

https://github.com/pymedphys/pymedphys/blob/main/lib/pymedphys/_dicom/ct/extend.py

with tests over at:

https://github.com/pymedphys/pymedphys/blob/main/lib/pymedphys/tests/dicom/test_extend_ct.py

It's heavily dependent on the `SliceLocation` parameter. And I have suspicions that there is a range of cases not accounted for. If that is the case I would prefer to test for those cases and raise errors, as opposed to supporting them at this time.

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py]: I had no idea about that. Thanks @sjswerdloff 🙂

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py]: Alright, so this PR now includes an additional `require_patient_orientation(ct_series, "HFS")` which should cover the cases I could envisage where this extend CT code will fail mightily.

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:127]: The root `extend` function also makes heavy use of `SliceLocation`. I shall refactor this out.

Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py]: Fixed in b43c97d11e51049a401ab37feaa5029b4e5c6d6f

**Comment by SimonBiggs** [comment]
> Prefer responses before approval

Hi @sjswerdloff,

I really appreciated this review. Thank you.

---

The PR did grow quite significantly. Let me know if you'd like me to split this one up.

Cheers,
Simon

File: lib/pymedphys/_dicom/ct/extend.py
Comment (sjswerdloff) [inline lib/pymedphys/_dicom/ct/extend.py:111]: what about the Series Instance UID?
You've generated a new Series, you weren't the owner of the original Series Instance UID (probably) so you need to put these in a different Series just for that reason alone, but more practically, if you re-use the same series UID, a lot of systems may put the old ones and these new ones together and try to load them as a single volume.
In general a CT series is a single volume.  You can get in to "but I only changed the data by using a different reconstruction smoothing kernel, they are the same acquired sinogram data", but it doesn't help in terms of making the data more interoperable.

**Review by sjswerdloff — CHANGES_REQUESTED** [review]
Series UID not being updated.
Trust me on this one.  Re-use of the Series UID when it shouldn't be can result in nasty issues.

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (sjswerdloff) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py]: Patient Position (0018,5100) is independent of Image Orientation Patient.
I have seen CT where the Image Orientation Patient was flipped around instead of the pixels (old Picker CT), so if Image Orientation Patient isn't what you would expect given Patient Positionn, things can get ugly.
I haven't seen that happen with HFS though, so given there aren't too many Picker CT floating around anymore...
(This isn't a ding on Picker... they had a right to do that... it just made life a lot harder for me...)

**Comment by sjswerdloff** [comment]
> > Prefer responses before approval
> 
> Hi @sjswerdloff,
> 
> I really appreciated this review. Thank you.
> 
> The PR did grow quite significantly. Let me know if you'd like me to split this one up.
> 
> Cheers,
> Simon

It grew because it *had* to.
I don't mind that, especially given I was able to review a fair amount and treat it as incremental.

**Comment by SimonBiggs** [comment]
Just a heads up, I won't look into these comments until Monday. I'll keep this one within work-time... given it's not something I do out of the love of my heart...

File: lib/pymedphys/_dicom/ct/extend.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/ct/extend.py]: Yeah, I shouldn't have even been making UIDs myself when `pydicom` does it appropriately. Fixed in 338e1c6f9216d609929b5eed71fbeb4be86b2947

File: lib/pymedphys/_dicom/ct/extend.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/ct/extend.py]: Are you okay with me vendoring this function? I didn't see any OnkoDICOM header at the top of that file, so I presumed that document came under the top-level license (source code is generally licensed on a per-file (document) basis, so it's best to have a header).

It does mean that a subset of PyMedPhys is under a different license, which isn't ideal. From looking at the blame it looks as if @Fincap and @sjswerdloff own the copyright to that file. Would you both be willing to have that function relicensed under the Apache-2.0 while retaining your copyright?

I underwent the change to fix this at 8d19ac2d086cd267baf059e31fc35127a43f1267

File: lib/pymedphys/_dicom/orientation.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/orientation.py]: Fixed? in cf92fce98916484b85b8a8520854332caaaebf1d

Quite keen for your advice on that one... I didn't include SITTING or FAFSIT as I used the following secondary source as a reference:

https://dicom.innolitics.com/ciods/ct-image/general-series/00185100

Happy to include SITTING and FAFSIT if it is appropriate...

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:173]: This is being sent to Monaco's DICOM receiver. Do think it is appropriate to put an ECHO in for that case?

File: lib/pymedphys/_dicom/ct/extend.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/ct/extend.py:108]: Mis-named the commit as I was interrupted mid coding, nevertheless, here's the commit that fixes this one:

add6a8aa491c09fbb84d652f20c00804abf1bbae 

> You've generated a new Series, you weren't the owner of the original Series Instance UID (probably) so you need to put these in a different Series just for that reason alone

This was certainly an issue. Thank you for picking this one up.

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:180]: Thanks for this one. Fixed in 13f804876cf3adc5c27303991b870fae2834d785

Let me know if that's what you were recommending.

**Comment by SimonBiggs** [comment]
@sjswerdloff just to let you know, don't feel obligated to get to the review of this one before you go on leave on Wednesday. I'm happy to leave this one hanging until you get back, I am also happy for it to be absolutely blocked until you give it the final tick of approval.

Cheers,
Simon

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py]: > Patient Position (0018,5100) is independent of Image Orientation Patient.

Yup, I shall make sure to verify both.

**Review by sjswerdloff — APPROVED** [review]
Nice

File: lib/pymedphys/_dicom/ct/extend.py
Comment (sjswerdloff) [reply lib/pymedphys/_dicom/ct/extend.py]: I'm fine with change in license for that excerpted code.
The license is from the top, and we/OnkoDICOM should explicitly incorporate  in individual files.

File: lib/pymedphys/_dicom/orientation.py
Comment (sjswerdloff) [reply lib/pymedphys/_dicom/orientation.py]: Innolitics is very useful but it is not authoritative.
The additional SITTING definition is elsewhere in the standard.

If the list is for CT only you are in good shape for a few years.  If it's for RT Ion Plan... You aren't supporting a meaningful fraction of real world clinical cases.

File: lib/pymedphys/_experimental/streamlit/apps/extend_ct.py
Comment (sjswerdloff) [reply lib/pymedphys/_experimental/streamlit/apps/extend_ct.py:181]: Should be useful if the SCP is not accessible for whatever reason

File: lib/pymedphys/_dicom/ct/extend.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/ct/extend.py]: Thanks @sjswerdloff, I underwent the relicense in the following commit:

69ba127b82753dfae9a63e62d9fc7cc47463d2bd

File: lib/pymedphys/_dicom/orientation.py
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/orientation.py]: > The additional SITTING definition is elsewhere in the standard.

Yup, found it https://dicom.innolitics.com/ciods/rt-image/rt-image/00185100 🙂 

> If the list is for CT only you are in good shape for a few years. If it's for RT Ion Plan... You aren't supporting a meaningful fraction of real world clinical cases.

I'm personally only supporting CT, but I think it's appropriate to add it. Fixed in 30f34ec53ba492cc1a2a9bfa823fd42b4e9522c2.

> Unless you are Varian, in which case they use FAFSIT ("FAce First SITting"). Sometimes.

I've opted to not include something non-standard for now. Should the day come that something non-standard is needed, maybe there can be a shim approach, or something similar.

**Comment by SimonBiggs** [comment]
> Nice

Thanks @sjswerdloff 🙂.

I'll go ahead and merge this PR. If you're not happy with the wording/approach to the re-licensing let me know and I can revert/adjust this PR.

**Pull request merged**
